### PR TITLE
Add release notes for incremental compilation improvements

### PR DESCRIPTION
**Pull request by melix**
What the title says

Fixes https://github.com/gradle/gradle-private/issues/3375

### CODE REVIEW

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -69,6 +69,26 @@ ADD RELEASE FEATURES ABOVE
...
+This means that if the compile task outputs were fetched from the build cache, then no incremental compilation was performed on the next build: the initial build after a fetch from the cache was fast, but the next one was always a full recompilation.
+In Gradle 7.1 this is no longer the case, as the result of the analysis is now stored in the build cache, meaning that the first compilation after fetching from the build cache will be incremental.
+
+The next improvement is about performance: incremental compilation analysis adds overhead to compilation, because Gradle needs to extract symbols from class files, analyze a transitive graph of dependencies to determine the consumers of a particular symbol, etc.
+In Gradle 7.1, we have significantly reduced this overhead, as well as reduced the cost of storage of the analysis.
+On the `gradle` project itself, we were able to make incremental compilation up to twice as fast!
+
+Last, because of the way the Java compiler works, previous Gradle releases were forced to perform a full recompilation as soon as _any_ constant was changed in an upstream dependency.
```
Comment (wolfs) [inline subprojects/docs/src/docs/release/notes.md:88]: For this change, we also need to update https://github.com/gradle/gradle/blob/6e7ba25c80638c3afff960f8923d861fd4ca88fb/subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc#incremental-java-compilation

It also seems that the limitation in `Known issues` about "If there is a mismatch in the package declaration and the directory structure of source files" has been already fixed, so we can remove it as well, right?

And maybe we should also remove the part about constants in the incremental Groovy compiler section: https://github.com/gradle/gradle/blob/6e7ba25c80638c3afff960f8923d861fd4ca88fb/subprojects/docs/src/docs/userguide/jvm/groovy_plugin.adoc#incremental-groovy-compilation

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -69,6 +69,26 @@ ADD RELEASE FEATURES ABOVE
...
+### Improved incremental compilation for Java
+
+The Java incremental compiler received substantial improvements in this release:
+
+1. incremental compilation analysis is now stored in the build cache
+2. incremental compilation analysis is faster, uses less memory and disk space
+3. incremental compiler is now aware of constant consumers
+
+In previous Gradle releases, incremental compilation analysis was only stored locally.
+This means that if the compile task outputs were fetched from the build cache, then no incremental compilation was performed on the next build: the initial build after a fetch from the cache was fast, but the next one was always a full recompilation.
```
Comment (wolfs) [inline subprojects/docs/src/docs/release/notes.md:81]: Do we mention this somewhere in the user manual?

**Review by wolfs — COMMENTED** [review]
The release notes look good. Though I think we need to update the user manual as well.

File: subprojects/docs/src/docs/release/notes.md (hunk 1 above)
Comment (melix) [reply subprojects/docs/src/docs/release/notes.md:88]: Fair point, let me do this.

File: subprojects/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -69,6 +69,26 @@ ADD RELEASE FEATURES ABOVE
 
 ## New features and usability improvements
 
+### Improved incremental compilation for Java
+
+The Java incremental compiler received substantial improvements in this release:
+
+1. incremental compilation analysis is now stored in the build cache
+2. incremental compilation analysis is faster, uses less memory and disk space
+3. incremental compiler is now aware of constant consumers
```
Comment (pioterj) [inline subprojects/docs/src/docs/release/notes.md:78]: It wasn't immediately clear to me what "incremental compiler is now aware of constant consumers" is about.

How about "incrementation compilation now also works in the presence of constants"?

Comment (pioterj) [reply subprojects/docs/src/docs/release/notes.md:78]: Maybe there is a better way of phrase it but "aware" does not communicate the user value, and "constant consumers" is not entirely clear.

Comment (melix) [reply subprojects/docs/src/docs/release/notes.md:78]: mmm, I'm not convinced either. "Presence of constants" may sound like constants in java source files, when the reality is more about constants in upstream dependencies and their use in consumers.

Comment (melix) [reply subprojects/docs/src/docs/release/notes.md:78]: What about:
> 3. change of a constant in an upstream dependency doesn't trigger a full recompilation anymore

Comment (pioterj) [reply subprojects/docs/src/docs/release/notes.md:78]: Sounds good to me!

**Comment by melix** [comment]
@bot-gradle test and merge

**Comment by melix** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Pull request merged**